python/simulation.py

```diff
@@ -1266,16 +1273,20 @@ def _run_sources_until(self, cond, step_funcs):
         if self.fields is None:
             self.init_sim()
 
-        ts = self.fields.last_source_time()
+        if not isinstance(cond, list):
+            cond = [cond]
 
-        if isinstance(cond, numbers.Number):
-            new_cond = (ts - self.round_time()) + cond
-        else:
-            def f(sim):
-                return cond(sim) and sim.round_time() >= ts
-            new_cond = f
+        ts = self.fields.last_source_time()
+        new_conds = []
+        for i in range(len(cond)):
+            if isinstance(cond[i], numbers.Number):
+                new_conds.append((ts - self.round_time()) + cond[i])
+            else:
+                def f(sim):
+                    return cond[i](sim) and sim.round_time() >= ts
+                new_conds.append(f)
```

Comment:
It doesn't seem like we need this logic anymore.  Can't we just do
```
cond.append(ts)
```
to append `ts` to the other conditions?

Reply:
I don't think that works. Wouldn't It ignore other stopping conditions? E.g., if we have

```py
sim.run(until_after_sources=mp.stop_when_fields_decayed(...))
```

it would stop after time `ts` whether or not the fields have decayed.

Reply:
Why can't we just add it to the list of stopping conditions, rather than modifying the existing conditions, since the new code stops if any of the conditions are true?

Reply:
What you're saying is true for arguments to `until`, but for arguments to `until_after_sources` a stopping condition implies that 2 things must be true:

1. The stopping condition function returns `True`
2. The time is after the "last source time".

So, in a case like this

```py
sim.run(until=mp.stop_after_walltime(10))
```
the simulation should stop after 10 seconds, no matter what. But I would expect this

```py
sim.run(until_after_sources=mp.stop_after_walltime(10))
```

to stop after 10 seconds only if the last source time has been hit. If it takes 20 seconds to hit the last source time, then it should not stop after only 10 seconds, but continue the full 20 seconds.

Am I misunderstanding what `until_after_sources` is supposed to do?

Reply:
Ah, good point.  Another way to put it is that `until_after_sources` is *and* whereas our list of conditions is joined with *or*.